In [13]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd
import nltk as nk

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# machine learning
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier

# Accessing Files and Folders
import os

# nltk.download()

In [24]:
# json_data = open("data.json").read()
# data = json.loads(json_data)

# pd.read_json(open("data.json", "r", encoding="utf8"))
# basePath = os.path.dirname(os.path.abspath("data.json"))
# df = pd.read_json(basePath + "/data.json", orient = 'records', dtype={"A":str, "B":list})

data_df = pd.read_json('/home/airwolf/Data science Projects/Data Science Lab/data20k.json', lines=True)


In [25]:
data_20k = data_df[:20000]
data_20k

authors    doc_type  \
0      [{'name': 'Peter Kostelnik', 'id': '2702511795...               
1      [{'name': 'Fabio Martinelli', 'id': '210743870...  Conference   
2           [{'name': 'Jerzy Mycka', 'id': '263067851'}]     Journal   
3      [{'name': 'Emdad Hossain', 'id': '2017661848',...  Conference   
4      [{'name': 'Pavel Kordík', 'id': '419063071', '...  Conference   
5      [{'name': 'George Dimitoglou', 'id': '20572666...     Journal   
6      [{'name': 'Colin Gabriel Onita', 'id': '196574...  Conference   
7        [{'name': 'Sandiway Fong', 'id': '2260246652'}]               
8      [{'name': 'Frédéric Bimbot', 'id': '737579779'...  Conference   
9      [{'name': 'Jinfeng Huang', 'id': '2116595219'}...  Conference   
10     [{'name': 'Elena Paslaru Bontas', 'id': '34092...  Conference   
11     [{'name': 'Sunzeng Cai', 'id': '2703866035', '...  Conference   
12      [{'name': 'Stephen Canipe', 'id': '2398078686'}]               
13     [{'name': 'Brendan J. Florio', 'id': '19948004...     Journal   
14     [{'name': 'Iain A. Stewart', 'id': '2107301613'}]     Journal   
15     [{'name': 'Sven-Erik Heun', 'id': '2696130382'...               
16     [{'name': 'Lucas Rodrigo Raso Mattos', 'id': '...               
17     [{'name': 'Milica Milutinovic', 'id': '2086708...  Conference   
18     [{'name': 'Michael Döhler', 'id': '1979564641'...               
19     [{'name': 'Christos Kalloniatis', 'id': '29618...  Conference   
20     [{'name': 'Chun-Ching Chen', 'id': '2108643440...  Conference   
21     [{'name': 'Matilde Santos', 'id': '2172096080'...  Conference   
22     [{'name': 'Eun-Kyeong Kwon', 'id': '2169385848...     Journal   
23     [{'name': 'Dae Sung Lee', 'id': '2609442339', ...  Conference   
24     [{'name': 'Li Zhuo', 'id': '2116670941', 'org'...     Journal   
25     [{'name': 'Fabrizio Costa', 'id': '2149327411'...               
26     [{'name': 'Charles H. Gartside', 'id': '265516...  Conference   
27         [{'name': 'Sven Heßler', 'id': '2113778294'}]               
28     [{'name': 'William Wu', 'id': '2162477007', 'o...  Conference   
29     [{'name': 'Stephen Charles Darden', 'id': '241...  Conference   
...                                                  ...         ...   
19970  [{'name': 'Munindar P. Singh', 'id': '23066554...     Journal   
19971  [{'name': 'Jayachandar Kodali', 'id': '2711142...  Conference   
19972  [{'name': 'Hakmook Kang', 'id': '2127279955', ...     Journal   
19973  [{'name': 'Paolo Liberatore', 'id': '703137194...               
19974  [{'name': 'Stanislav Harizanov', 'id': '223134...  Conference   
19975  [{'name': 'Brian Jenkins', 'id': '2044801297'}...     Journal   
19976  [{'name': 'Karel De Vogeleer', 'id': '26080830...     Journal   
19977  [{'name': 'Sébastien Briais', 'id': '338127684...  Conference   
19978  [{'name': 'Martine Toda', 'id': '2165152444'},...  Conference   
19979      [{'name': 'Johann Bizer', 'id': '132431757'}]     Journal   
19980  [{'name': 'Richard Chaumont', 'id': '250948636...  Conference   
19981  [{'name': 'Kristal K. Trejo', 'id': '202523575...     Journal   
19982  [{'name': 'Shanshan Han', 'id': '2713793848', ...  Conference   
19983  [{'name': 'Zoltán Ésik', 'id': '2021889967', '...  Conference   
19984     [{'name': 'Otto Spaniol', 'id': '2673365069'}]               
19985  [{'name': 'Ashok K. Thareja', 'id': '210186144...               
19986  [{'name': 'Dário Augusto Borges Oliveira', 'id...  Conference   
19987    [{'name': 'Norbert Noury', 'id': '2615743020'}]               
19988  [{'name': 'Victoria Medina', 'id': '2342303645...  Conference   
19989  [{'name': 'Cesar Guzman', 'id': '2100564791'},...     Journal   
19990  [{'name': 'Dawid Wojcieszak', 'id': '257550049...  Conference   
19991    [{'name': 'Dirk Stelling', 'id': '2716797248'}]               
19992  [{'name': 'Jacob Kowalewski', 'id': '267538465...               
19993  [{'name': 'Jan Vanus', 'id': '2311478965', 'or...               
19994  [{'name': 

####  Which features are available in the dataset

In [26]:
print(data_20k.columns.values)

['authors' 'doc_type' 'doi' 'fos' 'id' 'indexed_abstract' 'issue'
 'n_citation' 'page_end' 'page_start' 'publisher' 'references' 'title'
 'venue' 'volume' 'year']


In [27]:
# preview the data
data_20k.head()

authors    doc_type  \
0  [{'name': 'Peter Kostelnik', 'id': '2702511795...               
1  [{'name': 'Fabio Martinelli', 'id': '210743870...  Conference   
2       [{'name': 'Jerzy Mycka', 'id': '263067851'}]     Journal   
3  [{'name': 'Emdad Hossain', 'id': '2017661848',...  Conference   
4  [{'name': 'Pavel Kordík', 'id': '419063071', '...  Conference   

                            doi  \
0                           NaN   
1   10.1007/978-3-319-03491-1_5   
2                           NaN   
3  10.1007/978-3-642-33065-0_34   
4  10.1007/978-3-540-87559-8_89   

                                                 fos          id  \
0  [{'name': 'Lernaean Hydra', 'w': 0.4178039}, {...   100001334   
1  [{'name': 'Trusted Computing', 'w': 0.63148589...  1000018889   
2  [{'name': 'Discrete mathematics', 'w': 0.47368...  1000022707   
3  [{'name': 'Biometrics', 'w': 0.529778063000000...   100004108   
4  [{'name': 'Pattern recognition', 'w': 0.453429...    10000571   

                                    indexed_abstract issue  n_citation  \
0                                                NaN                 2   
1  {'IndexLength': 173, 'InvertedIndex': {'Both':...                 2   
2  {'IndexLength': 49, 'InvertedIndex': {'A': [0]...                 0   
3  {'IndexLength': 82, 'InvertedIndex': {'In': [0...                 0   
4  {'IndexLength': 171, 'InvertedIndex': {'Comple...                 5   

  page_end page_start                     publisher  \
0       46         43                                 
1       84         70                Springer, Cham   
2       29         19                                 
3      328        319  Springer, Berlin, Heidelberg   
4      868        859  Springer, Berlin, Heidelberg   

                                          references  \
0                                                NaN   
1  [94181602, 1504669610, 1542792105, 1639158619,...   
2                           [1972178849, 2069792094]   
3  [1578000111, 2120433720, 2136461127, 213893135...   
4  [291899685, 1964166287, 2135293965, 2146842127...   

                                               title  \
0  Ontologies in HYDRA - Middleware for Ambient I...   
1  Remote Policy Enforcement for Trusted Applicat...   
2  A SIMPLE OBSERVATION REGARDING ITERATIONS OF F...   
3  Gait based human identity recognition from mul...   
4  The GAME Algorithm Applied to Complex Fraction...   

                                               venue volume  year  
0                                    {'raw': 'AMIF'}         2009  
1  {'raw': 'international conference on trusted s...         2013  
2  {'raw': 'Reports on Mathematical Logic', 'id':...     44  2009  
3  {'raw': 'international conference on algorithm...         2012  
4  {'raw': 'international conference on artificia...         2008

In [18]:
# preview the tail
data_20k.tail()


authors    doc_type  \
19995  [{'name': 'Mohammadreza Mousavizadeh', 'id': '...  Conference   
19996  [{'name': 'Alberto Martelli', 'id': '259357277...               
19997  [{'name': 'Stephane Cavailles', 'id': '2398748...     Journal   
19998  [{'name': 'Elena Zaitseva', 'id': '2617929609'...               
19999  [{'name': 'Dirk Van Gucht', 'id': '2143976917'...               

                                doi  \
19995                           NaN   
19996                           NaN   
19997                           NaN   
19998  10.1007/978-3-319-19147-8_14   
19999                           NaN   

                                                     fos          id  \
19995  [{'name': 'Support vector machine', 'w': 0.415...  1119550597   
19996  [{'name': 'Discrete mathematics', 'w': 0.45981...   111955572   
19997  [{'name': 'Natural language processing', 'w': ...   111955886   
19998  [{'name': 'Human error', 'w': 0.53653359999999...  1119562032   
19999  [{'name': 'Relational calculus', 'w': 0.603273...    11195638   

                                        indexed_abstract issue  n_citation  \
19995  {'IndexLength': 147, 'InvertedIndex': {'Online...                 2   
19996                                                NaN                54   
19997                                                NaN                 0   
19998  {'IndexLength': 140, 'InvertedIndex': {'A': [0...                 2   
19999                                                NaN                 3   

      page_end page_start                                   publisher  \
19995                      Americas Conference on Information Systems   
19996      186        180                                               
19997       36         34                                               
19998      251        229                              Springer, Cham   
19999                                                                   

                                              references  \
19995  [128340387, 1999615887, 2010844754, 2119037484...   
19996                                                NaN   
19997                                                NaN   
19998  [653188647, 1174784412, 1975082752, 1981473248...   
19999                                                NaN   

                                                   title  \
19995  Antecedents of Online Customers Reviews’ Helpf...   
19996  An Algorithm for Unification in Equational The...   
19997  Principes Formels de la FUSION SYMBOLIQUE pour...   
19998  New Methods for the Reliability Analysis of He...   
19999  High Level Data Manipulation Languages for Unn...   

                                                   venue volume  year  
19995  {'raw': 'americas conference on information sy...         2015  
19996                                     {'raw': 'SLP'}         1986  
19997  {'raw': 'Applied Informatics', 'id': '25047507...         1994  
19998  {'raw': 'Applications of Computational Intelli...         2016  
19999      {'raw': 'XP7.52 Workshop on Database Theory'}         1986

#### the distribution of numerical feature values across the samples

In [28]:
data_20k.describe()

id    n_citation          year
count  2.000000e+04  20000.000000  20000.000000
mean   2.700255e+08      5.963050   2005.794500
std    3.657314e+08     24.446516      8.960792
min    1.091000e+03      0.000000   1946.000000
25%    1.027012e+08      0.000000   2002.000000
50%    1.068583e+08      1.000000   2008.000000
75%    1.108712e+08      5.000000   2013.000000
max    1.119562e+09   2020.000000   2018.000000

#### papers taken from year 25% are of 2002....... 50%=,  75%=.,........... max=....... min=..........

In [21]:
data_20k.describe(include=np.object)

authors    doc_type  \
count                                           20000       20000   
unique                                          19903           5   
top     [{'name': 'Johann Bizer', 'id': '132431757'}]  Conference   
freq                                                6       10926   

                            doi  \
count                      8870   
unique                     8870   
top     10.1145/2740908.2742574   
freq                          1   

                                                      fos  \
count                                               19501   
unique                                              19501   
top     [{'name': 'Artificial intelligence', 'w': 0.0}...   
freq                                                    1   

                                         indexed_abstract  issue page_end  \
count                                               12481  20000    20000   
unique                                              12481     67     1796   
top     {'IndexLength': 106, 'InvertedIndex': {'Homomo...                   
freq                                                    1  18302     3408   

       page_start publisher references  \
count       20000     20000      12704   
unique       1828       687      11733   
top                                 []   
freq         2508      8892        964   

                                                    title  \
count                                               20000   
unique                                              20000   
top     Electrocardiogram (ECG): A New Burgeoning Util...   
freq                                                    1   

                                                    venue volume  
count                                               19403  20000  
unique                                               3630    513  
top     {'raw': 'conference of the international speec...         
freq                                                  632  15930

### for a data of 20k
total venues 19403 and distinct venue are 3630

total reference are 12704 and distinct references are 11733

authors = 20000 and distinct 19903

distinct publishers = 687 and total = 20k

In [20]:
data_20k[:10]

authors    doc_type  \
0  [{'name': 'Peter Kostelnik', 'id': '2702511795...               
1  [{'name': 'Fabio Martinelli', 'id': '210743870...  Conference   
2       [{'name': 'Jerzy Mycka', 'id': '263067851'}]     Journal   
3  [{'name': 'Emdad Hossain', 'id': '2017661848',...  Conference   
4  [{'name': 'Pavel Kordík', 'id': '419063071', '...  Conference   
5  [{'name': 'George Dimitoglou', 'id': '20572666...     Journal   
6  [{'name': 'Colin Gabriel Onita', 'id': '196574...  Conference   
7    [{'name': 'Sandiway Fong', 'id': '2260246652'}]               
8  [{'name': 'Frédéric Bimbot', 'id': '737579779'...  Conference   
9  [{'name': 'Jinfeng Huang', 'id': '2116595219'}...  Conference   

                            doi  \
0                           NaN   
1   10.1007/978-3-319-03491-1_5   
2                           NaN   
3  10.1007/978-3-642-33065-0_34   
4  10.1007/978-3-540-87559-8_89   
5                           NaN   
6                           NaN   
7                           NaN   
8                           NaN   
9                           NaN   

                                                 fos          id  \
0  [{'name': 'Lernaean Hydra', 'w': 0.4178039}, {...   100001334   
1  [{'name': 'Trusted Computing', 'w': 0.63148589...  1000018889   
2  [{'name': 'Discrete mathematics', 'w': 0.47368...  1000022707   
3  [{'name': 'Biometrics', 'w': 0.529778063000000...   100004108   
4  [{'name': 'Pattern recognition', 'w': 0.453429...    10000571   
5  [{'name': 'Refinement', 'w': 0.570559442000000...   100007563   
6  [{'name': 'Continuance', 'w': 0.51218146100000...   100008278   
7  [{'name': 'Parsing', 'w': 0.5950031}, {'name':...   100008490   
8  [{'name': 'Usability', 'w': 0.5327128}, {'name...   100008599   
9  [{'name': 'Real-time computing', 'w': 0.461091...   100008749   

                                    indexed_abstract issue  n_citation  \
0                                                NaN                 2   
1  {'IndexLength': 173, 'InvertedIndex': {'Both':...                 2   
2  {'IndexLength': 49, 'InvertedIndex': {'A': [0]...                 0   
3  {'IndexLength': 82, 'InvertedIndex': {'In': [0...                 0   
4  {'IndexLength': 171, 'InvertedIndex': {'Comple...                 5   
5  {'IndexLength': 87, 'InvertedIndex': {'Negotia...                 1   
6  {'IndexLength': 126, 'InvertedIndex': {'IS': [...                 2   
7                                                NaN                 3   
8  {'IndexLength': 152, 'InvertedIndex': {'The': ...                10   
9                                                NaN                 2   

  page_end page_start                     publisher  \
0       46         43                                 
1       84         70                Springer, Cham   
2       29         19                                 
3      328        319  Springer, Berlin, Heidelberg   
4      868        859  Springer, Berlin, Heidelberg   
5      656        650                                 
6                 238                                 
7       60         43                                 
8      292        287                                 
9      278        264                                 

                                          references  \
0                                                NaN   
1  [94181602, 1504669610, 1542792105, 1639158619,...   
2                           [1972178849, 2069792094]   
3  [1578000111, 2120433720, 2136461127, 213893135...   
4  [291899685, 1964166287, 2135293965, 2146842127...   
5  [1578963809, 1612191182, 1964167056, 207843532...   
6  [116282327, 1496783311, 1565838829, 1570850622...   
7                                                NaN   
8  [1546801087, 1575951498, 2032085270, 212399132...   
9  [23142202, 1483005138, 1485432887, 1503973138,...   

                                               title  \
0  Ontologies in HYDRA - Middleware for Ambient I...   
1  Remote P

In [33]:
data_20k["authors"] // # data_20k.get("authors")


0        [{'name': 'Peter Kostelnik', 'id': '2702511795...
1        [{'name': 'Fabio Martinelli', 'id': '210743870...
2             [{'name': 'Jerzy Mycka', 'id': '263067851'}]
3        [{'name': 'Emdad Hossain', 'id': '2017661848',...
4        [{'name': 'Pavel Kordík', 'id': '419063071', '...
5        [{'name': 'George Dimitoglou', 'id': '20572666...
6        [{'name': 'Colin Gabriel Onita', 'id': '196574...
7          [{'name': 'Sandiway Fong', 'id': '2260246652'}]
8        [{'name': 'Frédéric Bimbot', 'id': '737579779'...
9        [{'name': 'Jinfeng Huang', 'id': '2116595219'}...
10       [{'name': 'Elena Paslaru Bontas', 'id': '34092...
11       [{'name': 'Sunzeng Cai', 'id': '2703866035', '...
12        [{'name': 'Stephen Canipe', 'id': '2398078686'}]
13       [{'name': 'Brendan J. Florio', 'id': '19948004...
14       [{'name': 'Iain A. Stewart', 'id': '2107301613'}]
15       [{'name': 'Sven-Erik Heun', 'id': '2696130382'...
16       [{'name': 'Lucas Rodrigo Raso Mattos', 'id': '.

In [77]:
# all_authors = [value["name"] for entry in data_20k["authors"] for value in entry]


['Peter Kostelnik',
 'Martin Sarnovsky',
 'Jan Hreno',
 'Fabio Martinelli',
 'Ilaria Matteucci',
 'Andrea Saracino',
 'Daniele Sgandurra',
 'Jerzy Mycka',
 'Emdad Hossain',
 'Girija Chetty',
 'Pavel Kordík',
 'Václav Křemen',
 'Lenka Lhotska',
 'George Dimitoglou',
 'Okan Duzyol',
 'Lawrence N. Owusu',
 'Colin Gabriel Onita',
 'Jasbir Singh Dhaliwal',
 'Sandiway Fong',
 'Frédéric Bimbot',
 'Emmanuel Deruty',
 'Gabriel Sargent',
 'Emmanuel Vincent',
 'Jinfeng Huang',
 'Jeroen Voeten',
 'Andre Ventevogel',
 'Elena Paslaru Bontas',
 'Sunzeng Cai',
 'Haiping Jiang',
 'Hua Qian',
 'Weidong Xiang',
 'Stephen Canipe',
 'Brendan J. Florio',
 'Michael Vynnycky',
 'Sarah L. Mitchell',
 "S. B. G. O'Brien",
 'Iain A. Stewart',
 'Sven-Erik Heun',
 'Thomas Heymann',
 'Jochen Schneider',
 'Fabian Schuster',
 'Gerald Spindler',
 'Lucas Rodrigo Raso Mattos',
 'Vijayaraghavan Varadharajan',
 'Rajarathnam Nallusamy',
 'Milica Milutinovic',
 'Vincent Naessens',
 'Bart De Decker',
 'Michael Döhler',
 'Xuan

In [63]:
for index in range(len(authorsEntry)):
    for key in authorsEntry[index]:
        print(authorsEntry[index][key])

Peter Kostelnik
2702511795
Martin Sarnovsky
2041014688
Jan Hreno
2398560122


'Peter Kostelnik'